### PR TITLE
Build and publish the docker image with Github actions

### LINKED ISSUE #1033: Move Docker build to Github actions
With the current limitations for docker builds on docker hub it's necessary to move the logic to github actions for good.

### PR DESCRIPTION
**Pull request by MichaelsJP**
…or pull requests. need to be removed before final merge

### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added: Build and publish docker image through github actions.
- Reason for change: Policy change of docker hub.

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

**Comment by MichaelsJP** [comment]
- [x] Docker build
- [x] Docker publish

**Comment by MichaelsJP** [comment]
still needed:
```
secrets.DOCKER_USERNAME
secrets.DOCKER_TOKEN
```

File: .github/workflows/build-publish.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+name: Build and publish Docker image with version
+on:
+  pull_request:
+    branches: [ master ]
+  push:
+    branches: [ master ]
+    tags:
+      - '*.*.*'
```
Comment (TheGreatRefrigerator) [inline .github/workflows/build-publish.yml]: I would keep array formats consistent with either
```suggestion
  pull_request:
    branches: [ master ]
  push:
    branches: [ master ]
    tags: ['*.*.*']
```
or
```suggestion
  pull_request:
    branches:
    - master
  push:
    branches:
    - master
    tags:
    - '*.*.*'
```

File: .github/workflows/build-publish.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      - name: Prepare
+        id: prepare
+        run: |
+          function test_version() {
+            curl -s -S "https://registry.hub.docker.com/v2/repositories/openrouteservice/openrouteservice/tags/?page_size=1024" |
+              sed -e 's/,/,\n/g' -e 's/\[/\[\n/g' |
+              grep '"name"' |
+              awk -F\" '{print $4;}' |
+              sort -fu
+          }
```
Comment (TheGreatRefrigerator) [inline .github/workflows/build-publish.yml]: A short comment on what exactly this one does would be great 😅
I think it shuld return a sorted list of the current tags (latest on top or below?)
But when i run this, it just returns 'next'. Is that intended?

File: .github/workflows/build-publish.yml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+          TAGS_VERSION="--tag ${DOCKER_IMAGE}:${DOCKER_VERSION}"
+          TAGS_LATEST="--tag ${DOCKER_IMAGE}:latest"
+          TAGS_NIGHTLY="--tag ${DOCKER_IMAGE}:nightly"
+
+          if [[ $CURRENT_VERSIONS =~ $DOCKER_VERSION ]]; then
+            echo "Version: $DOCKER_VERSION present or latest. Skipping it!"
+            BUILD_VERSION=false
+          else
+            echo "Version $DOCKER_VERSION not present and not latest. Building it!"
```
Comment (TheGreatRefrigerator) [inline .github/workflows/build-publish.yml]: I would rename this to IMAGE_VERSION or a more descriptive name, as this is not the docker version
```suggestion
          IMAGE_VERSION=${{ steps.previoustag.outputs.tag }}
          DOCKER_PLATFORMS=linux/amd64

          TAGS_VERSION="--tag ${DOCKER_IMAGE}:${IMAGE_VERSION}"
          TAGS_LATEST="--tag ${DOCKER_IMAGE}:latest"
          TAGS_NIGHTLY="--tag ${DOCKER_IMAGE}:nightly"

          if [[ $CURRENT_VERSIONS =~ $IMAGE_VERSION ]]; then
            echo "Image version: $IMAGE_VERSION present or latest. Skipping it!"
            BUILD_VERSION=false
          else
            echo "Image version $IMAGE_VERSION not present and not latest. Building it!"
```

**Review by TheGreatRefrigerator — COMMENTED** [review]
A few cents from my side :)

File: .github/workflows/build-publish.yml (hunk 1 above)
Comment (takb) [reply .github/workflows/build-publish.yml]: I agree. Also, does this work with our release tags having the format `vX.X.X`?

File: .github/workflows/build-publish.yml (hunk 2 above)
Comment (takb) [reply .github/workflows/build-publish.yml]: this seems to work fine on my machine and in the action's executions.

**Review by takb — APPROVED** [review]
looks good

**Pull request merged**